### ISSUE TITLE
Trouble cifar10 training example. - does not match CPU-only

### ISSUE THREAD

**Issue by Mandrewoid** [ROCmSoftwarePlatform/tensorflow#42]
Note: I am using python 3 version. Edit: Symptoms same in python2.7 version
Following the cifar10 example from https://github.com/ROCmSoftwarePlatform/tensorflow/blob/rocm-v1/rocm_docs/tensorflow-quickstart.md

It runs without any error messages, but it does not converge as it does with CPU-only.
For instance:
export HIP_VISIBLE_DEVICES=''
Training messages go like this:

2018-04-12 16:38:29.171171: step 9820, loss = 0.90 (752.3 examples/sec; 0.170 sec/batch)
2018-04-12 16:38:30.979742: step 9830, loss = 0.90 (707.7 examples/sec; 0.181 sec/batch)
2018-04-12 16:38:32.697638: step 9840, loss = 0.91 (745.1 examples/sec; 0.172 sec/batch)

Export HIP_VISIBLE_DEVICES=0
Training messages are now like this:

2018-04-12 16:46:37.732010: step 11600, loss = 1.01 (5222.9 examples/sec; 0.025 sec/batch)
2018-04-12 16:46:37.863226: step 11610, loss = 0.89 (9754.5 examples/sec; 0.013 sec/batch)
2018-04-12 16:46:37.987283: step 11620, loss = 0.92 (10318.4 examples/sec; 0.012 sec/batch)
2018-04-12 16:46:38.113692: step 11630, loss = 547.68 (10125.4 examples/sec; 0.013 sec/batch)
2018-04-12 16:46:38.235213: step 11640, loss = 547.89 (10533.3 examples/sec; 0.012 sec/batch)
2018-04-12 16:46:38.359570: step 11650, loss = 0.67 (10292.9 examples/sec; 0.012 sec/batch)

Any ideas what is going on here? Loss suddenly spikes to over 500, every 2 seconds this happens.

**Comment by whchung** [comment]
@Mandrewoid the "blips" in loss value is known for awhile, and it's due to potential issue inside the GPU kernel for `l2 loss` op. An experimental commit https://github.com/ROCmSoftwarePlatform/tensorflow/commit/321a784422acdac18ef38e0d98246174891896f8 is available but it's not merged to the main branch yet. I'm actually working on this issue as we speak.

**Comment by whchung** [comment]
+ @parallelo @b-sumner @scchan @dagamayank 

The issue seems to be in 1. compiler. Synchronization between `l2 loss` ops is lost due to compiler optimizations. The culprit seems to be in `https://github.com/RadeonOpenCompute/ROCm-Device-Libs` I myself wrote sometime ago which used incorrect LLVM function attributes. 

A fix is under verification now and hopefully it can be merged into the upcoming ROCm 1.7.x or 1.8.0 release.

**Comment by whchung** [comment]
Beside the issue in 1. compiler, 3. host kernel call site need also be improved. PR #44 tries to address this issue.

Notice in the very near future operators which depends on `Eigen` reduction such as `l2 loss` affected in this example would be switched to use `rocPRIM` ( https://github.com/ROCmSoftwarePlatform/rocprim ) which shouldn't run into this issue.

**Comment by Mandrewoid** [comment]
@whchung perhaps https://github.com/ROCmSoftwarePlatform/rocprim is a private repo. Thank link is 404 for me

**Comment by whchung** [comment]
@Mandrewoid yep it's something to be made public really soon. we have some PRs also pending to the open of the repo.

**Comment by VincentSC** [comment]
It was indeed made public 2 days after

**Comment by whchung** [comment]
Closing this issue as ROCm 1.7.2 has been released and PIP packages for ROCm 1.7.2 were also released.
